tests/functional/test_install.py

```diff
@@ -1527,3 +1528,99 @@ def test_target_install_ignores_distutils_config_install_prefix(script):
+        ' '.join(new_command)) in result.stderr, str(result)
+
+
+@pytest.mark.skipif("sys.platform != 'win32'")
+def test_protect_pip_from_modification_via_deps_on_windows(script, with_wheel):
+    """
+    Test ``pip install pkga`` is raised an error on Windows
+    if `pkga` implicitly tries to upgrade pip.
+    """
+    # Make a wheel for pkga which requires pip
+    script.scratch_path.joinpath('pkga').mkdir()
+    pkga_path = script.scratch_path / 'pkga'
+    pkga_path.joinpath('setup.py').write_text(textwrap.dedent("""
+        from setuptools import setup
+        setup(name='pkga',
+              version='0.1',
+              install_requires = ["pip<{}"])
+    """.format(pip.__version__)))
+    script.run(
+        'python', 'setup.py', 'bdist_wheel', '--universal', cwd=pkga_path
+    )
+
+    # Make sure pip install pkga is raised error
+    pkga_wheel_path = './pkga/dist/pkga-0.1-py2.py3-none-any.whl'
+    command = ['pip', 'install', pkga_wheel_path]
+    result = script.run(*command, expect_error=True)
+    assert result.returncode != 0
+    new_command = [sys.executable, "-m"] + command
+    assert "To modify pip, please run the following command:\n{}".format(
+        " ".join(new_command)) in result.stderr, str(result)
+
+
+@pytest.mark.skipif("sys.platform != 'win32'")
+def test_protect_pip_from_modification_via_sub_deps_on_windows(
+    script, with_wheel
+):
+    """
+    Test ``pip install pkg`` is raised an error on Windows
+    if sub-dependencies of `pkg` implicitly tries to upgrade pip.
+    """
+    # Make a wheel for pkga which requires pip
+    script.scratch_path.joinpath('pkga').mkdir()
+    pkga_path = script.scratch_path / 'pkga'
+    pkga_path.joinpath('setup.py').write_text(textwrap.dedent("""
+        from setuptools import setup
+        setup(name='pkga',
+              version='0.1',
+              install_requires = ["pip<{}"])
+    """.format(pip.__version__)))
+    script.run(
+        'python', 'setup.py', 'bdist_wheel', '--universal', cwd=pkga_path
+    )
+
+    # Make a wheel for pkgb which requires pkga
+    script.scratch_path.joinpath('pkgb').mkdir()
+    pkgb_path = script.scratch_path / 'pkgb'
+    pkgb_path.joinpath('setup.py').write_text(textwrap.dedent("""
+        from setuptools import setup
+        setup(name='pkgb',
+              version='0.1',
+              install_requires = ["pkga"])
+    """))
+    script.run(
+        'python', 'setup.py', 'bdist_wheel', '--universal', cwd=pkgb_path
+    )
+
+    # Make sure pip install pkgb is raised an error
+    pkgb_wheel_path = './pkgb/dist/pkgb-0.1-py2.py3-none-any.whl'
+    command = [
+        'pip', 'install', pkgb_wheel_path, '--find-links', pkga_path / 'dist'
+    ]
+    result = script.run(*command, expect_error=True)
+    assert result.returncode != 0
+    new_command = [sys.executable, '-m'] + command
+    assert 'To modify pip, please run the following command:\n{}'.format(
```

Comment:
nit: Construct the message and assign it to a variable called `expected_message`. Then this check can be:

```py
assert expected_message in result.stderr, result
```

Reply:
Did you mean?
```
assert expected_message in result.stderr, str(result)
```

otherwise

```
assert expected_message in result.stderr, result
```

gives useless assertion message:

```
AssertionError: <tests.lib.TestPipResult object at 0x10fbeada0>
```

Reply:
> `assert expected_message in result.stderr, str(result)`

Yep yep. 🙈